java/ql/src/Security/CWE/CWE-378/TempDirHijackingVulnerability.ql

```diff
@@ -0,0 +1,141 @@
+/**
+ * @name Temporary Directory Hijacking Vulnerability disclosure
+ * @description Detect temporary directory hijack vulnerability
+ * @kind path-problem
+ * @problem.severity error
+ * @precision very-high
+ * @id java/temp-directory-hijacking
+ */
+
+import java
+import semmle.code.java.controlflow.Guards
+import semmle.code.java.dataflow.FlowSources
+import semmle.code.java.security.TempFileLib
+import DataFlow::PathGraph
+
+/**
+ * An expression that will create a directory without throwing an exception if a file/directory already exists.
+ */
+private predicate isNonThrowingDirectoryCreationExpression(Expr expr, MethodAccess creationCall) {
+  creationCall.getMethod() instanceof MethodFileCreatesDirs and creationCall.getQualifier() = expr
+}
+
+private class MethodFileDelete extends Method {
+  MethodFileDelete() {
+    getDeclaringType() instanceof TypeFile and
+    hasName("delete")
+  }
+}
+
+predicate isDeleteFileExpr(Expr expr) {
+  expr = any(MethodFileDelete m).getAReference().getQualifier()
+}
+
+private class TempDirHijackingToDeleteConfig extends TaintTracking::Configuration {
+  TempDirHijackingToDeleteConfig() { this = "TempDirHijackingToDeleteConfig" }
+
+  override predicate isSource(DataFlow::Node source) {
+    source.asExpr() =
+      any(MethodAccess ma |
+        ma.getMethod() instanceof MethodFileCreateTempFile and ma.getNumArgument() = 2
+      )
+    or
+    // TODO: Replace with getSystemProperty("java.io.tmpdir")
+    source.asExpr() =
+      any(MethodAccessSystemGetProperty maSgp |
+        maSgp.hasCompileTimeConstantGetPropertyName("java.io.tmpdir")
+      )
+  }
+
+  override predicate isAdditionalTaintStep(DataFlow::Node node1, DataFlow::Node node2) {
+    node2.asExpr() =
+      any(MethodAccess ma |
+        ma.getMethod() instanceof MethodFileCreateTempFile and ma.getArgument(2) = node1.asExpr()
```

Comment:
Is this taint step correct in case the source is already a `MethodFileCreateTempFile`?
It looks like this taint step is only intended for the case where `File.createTempFile` is called with `java.io.tmpdir` as parent directory.

(Though apparently the tests cover this and pass?)